posts/inside-rust/2020-06-24-lto-improvements.md

```diff
@@ -0,0 +1,89 @@
+
+* Object code, which is the result of code generation. This is used during regular linking.
+* [LLVM bitcode], which is a binary representation of LLVM's intermediate representation. This can be used for [Link Time Optimization] (LTO).
+
+LTO is an optimization technique that can perform whole-program analysis. It analyzes all of the bitcode from every library at once, and performs optimizations and code generation. `rustc` supports several forms of LTO:
+
+* Fat LTO. This performs "full" LTO, which can take a long time to complete and may require a significant amount of memory.
+* [Thin LTO]. This is a lightweight version of "fat" LTO that can achieve similar performance improvements while taking much less time to complete.
+* Thin-local LTO. By default, `rustc` will split a crate into multiple "codegen units" so that they can be processed in parallel by LLVM. But this prevents some optimizations as code is separated into different codegen units, and is handled independently. Thin-local LTO will perform thin LTO across the codegen units within a single crate, bringing back some optimizations that would otherwise be lost by the separation. This is `rustc`'s default behavior if opt-level is greater than 0.
+
+## What has changed
+
+Changes have been made to both `rustc` and Cargo to control which libraries should include object code and which should include bitcode based on the project's [profile] LTO settings. If the project is not using LTO, then Cargo will instruct `rustc` to not place bitcode in the rlib files, which should reduce the amount of disk space used. This may have a small improvement in performance since `rustc` no longer needs to compress and write out the bitcode.
+
+If the project is using LTO, then Cargo will instruct `rustc` to not place object code in the rlib files, avoiding the expensive code generation step. This should improve the build time when building from scratch, and reduce the amount of disk space used.
+
+Two `rustc` flags are now available to control how the rlib is constructed:
+
+* [`-C linker-plugin-lto`] causes `rustc` to only place bitcode in the `.o` files, and skips code generation. Cargo uses this when the rlib is only intended for use with LTO. This can also be used when doing cross-language LTO.
+* [`-C embed-bitcode=no`] causes `rustc` to avoid placing bitcode in the rlib altogether. Cargo uses this when LTO is not being used, which reduces some disk space usage.
+
+Additionally, the method in which bitcode is embedded in the rlib has changed. Previously, `rustc` would place compressed bitcode as a `.bc.z` file in the rlib archive. Now, the bitcode is placed as an uncompressed section within each `.o` [object file] in the rlib archive. This avoids a small performance hit for compressing the bitcode, and also matches the standard format used by clang.
```

Comment:
The performance effect here is a mixed bag. Sometimes it's a speedup because we avoid the compression cost. Sometimes it's a slowdown because more data must be written. The combined effect is hard to predict and depends on the program. Avoiding the compression certainly makes the compiler simpler, though.

Reply:
It's also probably worth pointing out that "standard format" is a bit of a stretch as we discovered because only a few linkers actually work with the default output, we had to frob with some global inline asm to get the sections to work out on all platforms.

In reality this "standard format" was only created for iOS where they ship bitcode by default but wanted a mode of shipping both for debugging and such. We're kinda just piggy-backing on it at this point. In any case the words written down here are correct, so I'll leave it up to you if you'd like to add more context!

Reply:
Thanks for the info!  Updated with a little extra information.